framework/components/AFeatureDiscovery/AFeatureDiscovery.js

```diff
@@ -0,0 +1,249 @@
+import "./AFeatureDiscovery.scss";
+
+const ScrollListener = (onScroll, onScrollStop) => {
+  let handle = null;
+  const onScrollHandler = () => {
+    onScroll();
+    if (handle) {
+      clearTimeout(handle);
+    }
+    handle = setTimeout(onScrollStop, 200); // 200 ms
+  };
+  window.addEventListener("scroll", onScrollHandler);
+  return () => {
+    clearTimeout(handle);
+    window.removeEventListener("scroll", onScrollHandler);
+  };
+};
+
+const AFeatureDiscovery = forwardRef(
+  (
+    {
+      children,
+      disableScrollIntoView = false,
+      steps,
+      currentStep = 0,
+      className: propsClassName,
+      open = false,
+      onPrev,
+      onNext,
+      onClose,
+      role = "alertdialog",
+      ...rest
+    },
+    ref
+  ) => {
+    let className = `a-feature-discovery`;
+    if (propsClassName) {
+      className += ` ${propsClassName}`;
+    }
+    let backdropClassName = "a-feature-discovery-backdrop";
+    if (open) {
+      backdropClassName += " a-feature-discovery-backdrop--active";
+      className += " a-feature-discovery--active";
+    }
+    const step = steps[currentStep];
+
+    const {appRef} = useContext(AAppContext);
+    const {targetSelector, placement} = step;
+    const targetRef = useRef();
+    targetRef.current =
+      appRef.current?.querySelector(targetSelector) || appRef.current;
+    const [scrolling, setScrolling] = useState();
+    const [scrollingSettled, setScrollingSettled] = useState(
+      !disableScrollIntoView
+    );
+
+    const stepRef = useRef(null);
+    const combinedRef = useCombinedRefs(ref, stepRef);
+
+    const keyDownEscHandler = useCallback(
+      (e) => {
+        if (e.keyCode === keyCodes.esc) {
+          onClose && onClose();
+        }
+      },
+      [onClose]
+    );
+
+    useEffect(() => {
+      const mountEl = appRef.current;
+      if (!mountEl) return;
+      mountEl.addEventListener("keydown", keyDownEscHandler);
+
+      return () => mountEl.removeEventListener("keydown", keyDownEscHandler);
+    }, [appRef, keyDownEscHandler]);
+
+    const scrollingSettledtimeout = useRef(),
+      scrollIntoViewTimeout = useRef();
+    useEffect(() => {
+      if (disableScrollIntoView || !targetRef.current || !open) {
+        return;
+      }
+      const clearTimeouts = () => {
+        if (scrollingSettledtimeout.current) {
+          clearTimeout(scrollingSettledtimeout.current);
+        }
+        if (scrollIntoViewTimeout.current) {
+          clearTimeout(scrollIntoViewTimeout.current);
+        }
+      };
+      clearTimeouts();
+      setScrollingSettled(false);
+      scrollingSettledtimeout.current = setTimeout(
+        () => setScrollingSettled(true),
+        200
+      );
+      scrollIntoViewTimeout.current = setTimeout(
+        () =>
+          targetRef.current.scrollIntoView({
+            behavior: "smooth",
+            block: "center"
+          }),
+        10
+      );
+
+      () => clearTimeouts();
```

Comment:
not sure what this line does

Reply:
Yeah but it's not executing anything. It's a function declaration, and it's not possible to call this function.